### PR TITLE
Rewrite using MsBuild APIs

### LINKED ISSUE #73: ProjInfo.Inspect.getProjectInfos fails for TargetFrameworks
**Describe the bug**

Using FSharpLint for projects with `TargetFramework` works fine, as soon as a project is using mutli-targetting with `TargetFrameworks` the linting fails.
I narrowed it down to `FSharpLint.Application.Lint.getProjectFileInfo` which throws an exception.

**To Reproduce**

Use [PR 403](https://github.com/fsprojects/FSharpLint/pull/403) of FSharpLint.

1. Build the project with `build.cmd`
Already on the console log you'll see the messages as in the "screenshot" below

2. Set a breakpoint in `Application\Lint.fs` and debug
(Branch `dotnet-proj-info` at https://github.com/milbrandt/FSharpLint.git)
A configured launch file is in this branch included (2 configurations, one with onw qwithout exception)
Set a breakpoint in `getProjectFileInfo` at `match getFscArgsResult with` (line 311)
This line fails, see screenshot.

**Expected behaviour**

Working for projects with `TargetFrameworks`, either a list or for the first target framework.

**Screenshots**

```
  FSharpLint.Console -> D:\_Repos\FSharpLint\src\FSharpLint.Console\bin\Release\netcoreapp3.1\dotnet-fsharplint.dll
  FSharpLint.Console -> D:\_Repos\FSharpLint\src\FSharpLint.Console\bin\Release\netcoreapp2.1\dotnet-fsharplint.dll
EXEC : FSharpLint error : Lint failed while analysing D:\_Repos\FSharpLint\src\FSharpLint.Console\FSharpLint.Console.fsproj. [D:\_Repos\FSharpLint\src\FSharpLint.Console\FSharpLint.Console.fsproj]
  Failed with: meow
  Stack trace:    at FSharpLint.Application.Lint.getProjectFileInfo(FSharpOption`1 releaseConfig, String projectFilePath) in C:\projects\fsharplint-231\src\FSharpLint.Core\Application\Lint.fs:line 335
     at FSharpLint.Application.Lint.lintProject$cont@422(OptionalLintParameters optionalParams, String projectFilePath, Unit unitVar) in C:\projects\fsharplint-231\src\FSharpLint.Core\Application\Lint.fs:line 466
     at FSharpLint.Console.Program.start(ParseResults`1 arguments) in C:\projects\fsharplint-231\src\FSharpLint.Console\Program.fs:line 120
```
projinfo-exception

**Environment (please complete the following information):**
 - OS: Win10
 - Ionide version: ionide-VSCode: FSharp 4.9.0
 - VSCode version: 1.44.2 (not relevant)
 - dotnet SDK version: 3.1.200
 - mono / .Net Framework version: netcoreapp3.1;netcoreapp2.1

### LINKED ISSUE #81: /p: DesignTimeBuild=true needed to get perf for project queries
Hi all

I using this library to get project properties like OutputPath, IsPackable etc. for projects in https://github.com/fsprojects/FSharp.Formatting/pull/528

I was surprised how long each call to get these properties is taking, I'm seeing 5-10sec runtime for each project.  It seems the projects are actually building, because there are processes `fsc.exe` executing, which surprises me, I can't understand why a build would be happening given we have `/t:_Inspect_GetProperties` (there is nothing unusual about the project files)

I guess I'm just wondering if this is expected or if it's driver error?

The basics of what I'm doing appears to be the same as FsAutoComplete, without the caching.  The `msbuild`/`dotnet build` calls being generated are:

```
args = ["msbuild";
 "E:\GitHub\dsyme\FSharp.Formatting\tests\FSharp.CodeFormat.Tests\FSharp.CodeFormat.Tests.fsproj";
 "/t:_Inspect_GetProperties";
 "/p:_Inspect_GetProperties_OutFile=C:\Users\dsyme\AppData\Local\Temp\tmp2C34.tmp.GetProperties.txt";
 "/p:CustomAfterMicrosoftCommonTargets=C:\Users\dsyme\AppData\Local\Temp\tmp2C44.tmp.proj-info.hook.targets";
 "/p:CustomAfterMicrosoftCommonCrossTargetingTargets=C:\Users\dsyme\AppData\Local\Temp\tmp2C44.tmp.proj-info.hook.targets";
 "/nologo"; "/verbosity:quiet"]
```
Here's the outline of my code - the key thing is `Dotnet.ProjInfo.Inspect.msbuild msbuildPath runCmd` with the above arguments
```
        let additionalInfo =
            [ "OutputType"
              "IsTestProject"
              "IsPackable"
            ]
        let gp () = Dotnet.ProjInfo.Inspect.getProperties (["TargetPath"] @ additionalInfo)

        let loggedMessages = System.Collections.Concurrent.ConcurrentQueue<string>()
        let runCmd exePath args =
           printfn "args = %A" args
           runProcess loggedMessages.Enqueue slnDir exePath (args |> String.concat " ")
        let msbuildPath = Dotnet.ProjInfo.Inspect.MSBuildExePath.DotnetMsbuild "dotnet"
        let msbuildExec = Dotnet.ProjInfo.Inspect.msbuild msbuildPath runCmd

        let result = file |> Dotnet.ProjInfo.Inspect.getProjectInfos loggedMessages.Enqueue msbuildExec [gp] []
```

### LINKED ISSUE #82: multi-line property values like Description cause errors
This is a note of an incompleteness: using `getProperties` with multi-line property values doesn't work correctly.  The error messages that result can be fairly obscure.

An example of a property value that causes problems: 
```
    <Description>
      The package is a collection of libraries that can be used for literate programming
      with F# (great for building documentation) and for generating library documentation
      from inline code comments. The key componments are Markdown parser, tools for formatting
      F# code snippets, including tool tip type information and a tool for generating
      documentation from library metadata.
    </Description>
```

### LINKED ISSUE #87: Proposal: Use MSBuild's Microsoft.Build.* libraries
Currently, `dotnet-proj-info` works with project files by invoking the command-line edition of MSBuild. This approach becomes unwidely when retrieving information from a project: a temporary target file is generated that specifies the names of the proeprties it needs, then MSBuild is invoked, writing the values of these properties to another temporary file. And this approach isn't perfect, resulting in bugs like #82. Even worse, pieces of the source code are MSBuild is being selectively piggybacked in a pretty brittle and clearly unsupported way, via the internal `Dotnet.ProjInfo.Helpers` project.

What I propose is to use the official MSBuild API, as offered through packages like [`Microsoft.Build`](https://nuget.org/packages/Microsoft.Build). The benefits would be enormous. The codebase will be simplified, the performance will be increased, and the project will follow established best practices.

### Caveats

This is gonna be a breaking change, more or less, but acceptable for me since we are still in version 0.x:

* __Support for .NET Standard will have to be removed.__ Only .NET Framework and Core will be supported. Personally I don't consider it a problem; I cannot imagine a .NET project analyzer that needs the ubiquity of platforms that .NET Stnadard can offer but .NET Core 2.x cannot. Besides, .NET 5's time draws near, reducing the framework fragmantation.

* We have to reconsider the way the project deals with MSBuild installations. Currently it looks for MSBuild executable _files_, but what we would need are _directories_ with the MSBuild assemblies. We would use the [`Microsoft.Build.Locator`](https://nuget.org/packages/Microsoft.Build.Locator) package [as recommended by Microsoft](https://docs.microsoft.com/visualstudio/msbuild/updating-an-existing-application#use-microsoftbuildlocator). It can find only MSBuild installations from .NET Core SDKs when run on .NET Core, and only from Visual Studio when run from .NET Framework.

  * What is more, the MSBuild locator's implementation relies on hooking assembly resolving events, meaning that we might not be able to use multiple versions of MSBuild per program invocation. Still it seems a pretty exotic scenario for me.

* To read solution files, we use [a package that was last updated three years ago](https://www.nuget.org/packages/Sln) that also piggybacks from MSBuild source code. Using the official solution parser API would also cause trouble because it cannot retrieve the files inside a solution folder, which means that until dotnet/msbuild#1708 is fixed (preferably by enricosada/Sln#3), we have to keep using the package we are already using.

---

I am waiting for your feedback on these proposals. Once some ambiguities (like my second bullet point) are addressed, I will start working on an implementation.

### PR DESCRIPTION
**Pull request by Krzysztof-Cieslak**
So this PR will contain a huge rewrite of the whole `dotnet-proj-info`. 

There are few goals of this rewrite:
1. Create a more maintainable code base - DPI has been evolving for years, with different lead developers which resulted in code base being... a bit of mess. At the moment it's kinda hard to follow the flow of what's going on, and making changes  is not the simplest thing
2. Use Microsoft.Build.* libraries - as discussed in #87
3. Port `Project System` library from FSAC - Project System is a high-level component that's responsible for things like watching fsproj files for changes, have nice API, and is easy to use. We've been asked to release it as a NuGet package, and in this context I think it'll be better to have it here than in FSAC repo. 
4. Port all FSI dependency resolution stuff we have in FSAC - it makes more sense to have it here, in the reusable library (I'll probably need @baronfel help here)
4. Drop support for old/verbose project files - maintaining support for them is not worth an effort. 

This PR right now introduces `Dotnet.ProjInfo.ProjectLoader` project which is using `Microsoft.Build.*` libraries to parse and evaluate project files. It outputs almost the same data structure as `Dotnet.ProjInfo.Workspace`. 
`Dotnet.ProjInfo.ProjectLoader` will replace both `Dotnet.ProjInfo` and `Dotnet.ProjInfo.Workspace` projects but right now I develop it side-by-side so I can easily copy-paste some code from old projects to the new one. 

Closes #87, closes #81, closes #82, closes #73

### CODE REVIEW

**Comment by Krzysztof-Cieslak** [comment]
`Dotnet.ProjInfo.ProjectLoader` is now again `Dotnet.ProjInfo`. old `Dotnet.ProjInfo`, `Dotnet.ProjInfo.Workspace` and `dotnet-proj` (CLI tool) are removed

**Comment by dsyme** [comment]
Great work!

**Comment by Krzysztof-Cieslak** [comment]
@dsyme I'll try to updater F#.Formatting on Monday or Tuesday

**Comment by dsyme** [comment]
> @dsyme I'll try to updater F#.Formatting on Monday or Tuesday

Cool ping me on slack or elsewhere if you need help (not twitter DM I don't get notifications on those)

**Comment by Krzysztof-Cieslak** [comment]
I'll actually probably wait few more days for initial bug reports from Ionide users and some bugfixes based on those.
